File: packages/engine/src/memory/shortMemory.ts
import { db } from 'db';
import type { ChatMessage } from 'types';
import { sanitizeTenantResponse } from '../graph/nodes/finish.node';
import { getEmbeddingModel } from '../llm/callLLMWithRetry';

export interface ShortMemoryMessage extends ChatMessage {
  role: 'user' | 'assistant' | 'system';
  content: string;
}

let lastGlobalTimestamp = 0;

function getMonotonicTimestamp(role?: 'user' | 'assistant' | 'system'): string {
  const now = Date.now();
  const roleOffset = role === 'assistant' ? 10 : role === 'system' ? 0 : 5;
  if (now <= lastGlobalTimestamp) {
    lastGlobalTimestamp += roleOffset > 0 ? roleOffset : 1;
  } else {
    lastGlobalTimestamp = now + roleOffset;
  }
  return new Date(lastGlobalTimestamp).toISOString();
}

export class ShortMemory {
  private threadId: string;
  private maxTurns: number;
  private businessId?: string;

  constructor(threadId: string, maxTurns = 10, businessId?: string) {
    this.threadId = threadId;
    this.maxTurns = maxTurns;
    if (businessId) {
      this.businessId = businessId;
    } else {
      const lower = threadId.toLowerCase();
      if (lower.includes('aurora')) this.businessId = 'aurora';
      else if (lower.includes('nike')) this.businessId = 'nike';
      else if (lower.includes('adidas')) this.businessId = 'adidas';
    }
  }

  async getMessages(): Promise<ShortMemoryMessage[]> {
    try {
      const messages = await db.getMessages(this.threadId);
      // Resolve thread businessId if available for sanitization
      let businessId = this.businessId || 'ecommerce';
      try {
        const { getDrizzle, threads } = require('db');
        const { eq } = require('drizzle-orm');
        const drizzle = getDrizzle();
        if (drizzle) {
          const threadRows = await drizzle.select().from(threads).where(eq(threads.id, this.threadId)).limit(1);
          if (threadRows[0]?.businessId) {
            businessId = threadRows[0].businessId;
          }
        }
      } catch {
        // Safe fallback
      }

      // Implement a sliding context window to fetch only the latest (maxTurns * 2) messages,
      // avoiding Context Window Bloat and reducing DB parsing and LLM token billing costs.
      const sliced = messages.slice(-this.maxTurns * 2);
      return sliced.map((msg: { role: string; content: string }) => ({
        role: msg.role as 'user' | 'assistant' | 'system',
        content: msg.role === 'assistant' ? sanitizeTenantResponse(msg.content, businessId) : msg.content,
      }));
    } catch (err) {
      console.error('[ShortMemory Error] Failed to get messages:', err);
      return [];
    }
  }

  async addMessage(role: 'user' | 'assistant' | 'system', content: string): Promise<void> {
    const id = crypto.randomUUID ? crypto.randomUUID() : Math.random().toString(36).substring(2, 15);
    const timestamp = getMonotonicTimestamp(role);
    const cleanContent = content !== undefined && content !== null ? String(content) : '';

    try {
      await db.createThread(this.threadId, '83d67d4e-104c-4325-8aa7-10d4389fc725', this.businessId);
      await db.addMessage({
        id,
        threadId: this.threadId,
        businessId: this.businessId,
        role,
        content: cleanContent,
        timestamp,
      } as any);
      console.log(
        `[ShortMemory] Added message for thread ${this.threadId}: [${role}] ${cleanContent.substring(0, 50)}`,
      );
    } catch (err) {
      console.error('[ShortMemory Error] Failed to add message:', err);
    }
  }

  async compress(messages: ShortMemoryMessage[]): Promise<string> {
    return 'Summary of compressed conversation history';
  }
}

File: services/engine-py/src/engine_py/memory/short_memory.py
"""短期记忆 — 镜像 packages/engine/src/memory/shortMemory.ts。

滑动窗口取最近 max_turns*2 条;assistant 内容经租户清洗后返回;
写入使用单调递增时间戳保证同毫秒消息的稳定排序。
"""

from __future__ import annotations

import datetime as _dt
import uuid

from sqlalchemy import case, select, text

from ..db import Message, Thread, get_session
from ..tenant import sanitize_tenant_response

# 与 TS ShortMemory.addMessage 保持一致的固定虚拟用户(线程自愈归属)
_FALLBACK_USER_ID = "83d67d4e-104c-4325-8aa7-10d4389fc725"

_last_global_timestamp_ms = 0


def _monotonic_timestamp(role: str | None) -> str:
    global _last_global_timestamp_ms
    now = int(_dt.datetime.now().timestamp() * 1000)
    role_offset = 10 if role == "assistant" else 0 if role == "system" else 5
    if now <= _last_global_timestamp_ms:
        _last_global_timestamp_ms += role_offset if role_offset > 0 else 1
    else:
        _last_global_timestamp_ms = now + role_offset
    return _dt.datetime.fromtimestamp(_last_global_timestamp_ms / 1000, tz=_dt.UTC).isoformat()


def _infer_business_id(thread_id: str, explicit: str | None = None) -> str:
    """镜像 db.createThread 的智能租户推导:显式值优先,否则按 threadId 关键词推断。"""
    if explicit and explicit != "ecommerce":
        return explicit
    lower = thread_id.lower()
    for key in ("aurora", "nike", "adidas"):
        if key in lower:
            return key
    return explicit or "ecommerce"


class ShortMemory:
    def __init__(self, thread_id: str, max_turns: int = 10, business_id: str | None = None) -> None:
        self.thread_id = thread_id
        self.max_turns = max_turns
        self.business_id = business_id or _infer_business_id(thread_id)

    async def get_messages(self) -> list[dict]:
        try:
            async with get_session() as session:
                rows = (
                    await session.execute(
                        select(Message)
                        .where(Message.thread_id == self.thread_id)
                        .order_by(
                            Message.timestamp,
                            case(
                                (Message.role == "system", 1),
                                (Message.role == "user", 2),
                                (Message.role == "assistant", 3),
                                else_=4,
                            ),
                            Message.id,
                        )
                    )
                ).scalars().all()

                thread_row = (
                    await session.execute(select(Thread).where(Thread.id == self.thread_id).limit(1))
                ).scalar_one_or_none()
                business_id = (thread_row.business_id if thread_row else None) or self.business_id or "ecommerce"

                sliced = list(rows)[-(self.max_turns * 2) :]
                return [
                    {
                        "role": m.role,
                        "content": (
                            sanitize_tenant_response(m.content, business_id)
                            if m.role == "assistant"
                            else m.content
                        ),
                        "cards": m.cards or None,
                    }
                    for m in sliced
                ]
        except Exception as err:
            print(f"[ShortMemory Error] Failed to get messages: {err}")
            return []

    async def add_message(
        self, role: str, content: str, cards: list | None = None
    ) -> None:
        clean_content = "" if content is None else str(content)
        try:
            async with get_session() as session:
                # 线程自愈(镜像 db.createThread:存在则刷新,不存在则建)
                await session.execute(
                    text(
                        'INSERT INTO threads (id, "user_id", "business_id", status, "created_at", "updated_at") '
                        "VALUES (:tid, :uid, :bid, 'active', NOW(), NOW()) "
                        'ON CONFLICT (id) DO UPDATE SET "updated_at" = NOW(), "business_id" = EXCLUDED."business_id"'
                    ).bindparams(tid=self.thread_id, uid=_FALLBACK_USER_ID, bid=self.business_id)
                )
                session.add(
                    Message(
                        id=str(uuid.uuid4()),
                        thread_id=self.thread_id,
                        business_id=self.business_id,
                        role=role,
                        content=clean_content,
                        cards=cards if cards else None,
                        timestamp=_monotonic_timestamp(role),
                    )
                )
                await session.commit()
        except Exception as err:
            print(f"[ShortMemory Error] Failed to add message: {err}")

    async def compress(self, messages: list[dict]) -> str:
        return "Summary of compressed conversation history"
